src/utils/display-observer.js

```diff
@@ -0,0 +1,248 @@
+    this.targetObserverCallbacks_.clear();
+    this.targetObservations_.clear();
+  }
+
+  /**
+   * @param {!Element} target
+   * @param {function(boolean)} callback
+   */
+  observe(target, callback) {
+    let callbacks = this.targetObserverCallbacks_.get(target);
+    if (!callbacks) {
+      callbacks = [];
+      this.targetObserverCallbacks_.set(target, callbacks);
+      this.observers_[DOC_OBSERVER].observe(target);
+      this.observers_[VIEWPORT_OBSERVER].observe(target);
+    }
+    if (pushIfNotExist(callbacks, callback)) {
+      if (this.targetObservations_.has(target)) {
+        // Notify the existing observation immediately.
+        setTimeout(() => {
+          const display = computeDisplay(
+            this.targetObservations_.get(target),
+            this.isDocVisible_
+          );
+          if (display != null) {
+            callCallbackNoInline(callback, display);
+          }
+        });
+      }
+    }
+  }
+
+  /**
+   * @param {!Element} target
+   * @param {function()} callback
+   */
+  unobserve(target, callback) {
+    const callbacks = this.targetObserverCallbacks_.get(target);
+    if (!callbacks) {
+      return;
+    }
+    removeItem(callbacks, callback);
+    if (callbacks.length == 0) {
+      this.targetObserverCallbacks_.delete(target);
+      this.targetObservations_.delete(target);
+      this.observers_[DOC_OBSERVER].unobserve(target);
+      this.observers_[VIEWPORT_OBSERVER].unobserve(target);
+    }
+  }
+
+  /** @private */
+  docVisibilityChanged_() {
+    this.targetObserverCallbacks_.forEach((callbacks, target) => {
+      const observations = this.targetObservations_.get(target);
+      const oldDisplay = computeDisplay(observations, !this.isDocVisible_);
+      const newDisplay = computeDisplay(observations, this.isDocVisible_);
+      if (newDisplay != null && newDisplay !== oldDisplay) {
+        for (let k = 0; k < callbacks.length; k++) {
+          callCallbackNoInline(callbacks[k], newDisplay);
+        }
+      }
+    });
+  }
+
+  /**
+   * @param {!Array<!IntersectionObserverEntry>} entries
+   * @param {number} observer
+   * @private
+   */
+  observed_(entries, observer) {
+    const seen = new Set();
+    for (let i = entries.length - 1; i >= 0; i--) {
+      const {target, isIntersecting} = entries[i];
+      if (seen.has(target)) {
+        continue;
+      }
+      seen.add(target);
+      const callbacks = this.targetObserverCallbacks_.get(target);
+      if (!callbacks) {
+        continue;
+      }
+      let observations = this.targetObservations_.get(target);
+      if (!observations) {
+        observations = [];
+        this.targetObservations_.set(target, observations);
+      }
+      const oldDisplay = computeDisplay(observations, this.isDocVisible_);
+      observations[observer] = isIntersecting;
+      const newDisplay = computeDisplay(observations, this.isDocVisible_);
+      if (newDisplay != null && newDisplay !== oldDisplay) {
```

Comment:
Nit: Can we extract this into a function? It's shared with `docVisibilityChanged_`